rcl/src/rcl/security.c

```diff
@@ -0,0 +1,189 @@
+{
+  const char * ros_security_enable = NULL;
+  const char * get_env_error_str = NULL;
+
+  RCL_CHECK_ARGUMENT_FOR_NULL(use_security, RCL_RET_INVALID_ARGUMENT);
+
+  get_env_error_str = rcutils_get_env(ROS_SECURITY_ENABLE_VAR_NAME, &ros_security_enable);
+  if (NULL != get_env_error_str) {
+    RCL_SET_ERROR_MSG_WITH_FORMAT_STRING(
+      "Error getting env var '" RCUTILS_STRINGIFY(ROS_SECURITY_ENABLE_VAR_NAME) "': %s\n",
+      get_env_error_str);
+    return RCL_RET_ERROR;
+  }
+
+  *use_security = (0 == strcmp(ros_security_enable, "true"));
+  return RCL_RET_OK;
+}
+
+rcl_ret_t
+rcl_get_enforcement_policy(rmw_security_enforcement_policy_t * policy)
+{
+  const char * ros_enforce_security = NULL;
+  const char * get_env_error_str = NULL;
+
+  RCL_CHECK_ARGUMENT_FOR_NULL(policy, RCL_RET_INVALID_ARGUMENT);
+
+  get_env_error_str = rcutils_get_env(ROS_SECURITY_STRATEGY_VAR_NAME, &ros_enforce_security);
+  if (NULL != get_env_error_str) {
+    RCL_SET_ERROR_MSG_WITH_FORMAT_STRING(
+      "Error getting env var '" RCUTILS_STRINGIFY(ROS_SECURITY_STRATEGY_VAR_NAME) "': %s\n",
+      get_env_error_str);
+    return RCL_RET_ERROR;
+  }
+
+  *policy = (0 == strcmp(ros_enforce_security, "Enforce")) ?
+    RMW_SECURITY_ENFORCEMENT_ENFORCE : RMW_SECURITY_ENFORCEMENT_PERMISSIVE;
+  return RCL_RET_OK;
+}
+
+char * exact_match_lookup(
+  const char * name,
+  const char * ros_secure_root_env,
+  const rcl_allocator_t * allocator)
+{
+  // Perform an exact match for the node/context's name in directory <root dir>/<namespace>.
+  char * secure_root = NULL;
+  // "/" case when root namespace is explicitly passed in
+  if (0 == strcmp(name, "/")) {
+    secure_root = rcutils_strdup(ros_secure_root_env, *allocator);
+  } else {
+    char * root_path = NULL;
+    // Get native path, ignore the leading forward slash
+    // TODO(ros2team): remove the hard-coded length, use the length of the root namespace instead
+    root_path = rcutils_to_native_path(name + 1, *allocator);
+    secure_root = rcutils_join_path(ros_secure_root_env, root_path, *allocator);
+    allocator->deallocate(root_path, allocator->state);
+  }
+  return secure_root;
+}
+
+char * rcl_get_secure_root(
+  const char * name,
+  const rcl_allocator_t * allocator)
+{
+  bool ros_secure_directory_override = true;
+
+  // find out if either of the configuration environment variables are set
+  const char * env_buf = NULL;
+  if (NULL == name) {
+    return NULL;
+  }
+
+  if (rcutils_get_env(ROS_SECURITY_DIRECTORY_OVERRIDE, &env_buf)) {
+    return NULL;
+  }
+  if (!env_buf) {
+    return NULL;
+  }
+  if (0 == strcmp("", env_buf)) {
+    // check root directory if override directory environment variable is empty
+    if (rcutils_get_env(ROS_SECURITY_ROOT_DIRECTORY_VAR_NAME, &env_buf)) {
+      return NULL;
+    }
+    if (!env_buf) {
+      return NULL;
+    }
+    if (0 == strcmp("", env_buf)) {
+      return NULL;  // environment variable was empty
+    } else {
```

Comment:
@ivanpauno nit: on need for an `else` clause

Reply:
See https://github.com/ros2/rcl/pull/515/commits/e3d5ebdf6094274f7f74d752ecce3fbadec84150.